syntax = "proto3";
package examples;
option go_package = "github.com/angzarr/angzarr/proto/examples;examples";
option java_package = "dev.angzarr.examples";
option java_multiple_files = true;
option csharp_namespace = "Angzarr.Examples";
option ruby_package = "Angzarr::Examples";

import "google/protobuf/timestamp.proto";

// =============================================================================
// Inventory Domain - manages stock levels and reservations
// =============================================================================

// Commands
message InitializeStock {
  string product_id = 1;
  int32 quantity = 2;
  int32 low_stock_threshold = 3;  // triggers LowStockAlert when available < threshold
}

message ReceiveStock {
  int32 quantity = 1;
  string reference = 2;  // e.g., purchase order number
}

message ReserveStock {
  int32 quantity = 1;
  string order_id = 2;
}

message ReleaseReservation {
  string order_id = 1;
}

message CommitReservation {
  string order_id = 1;
}

// Events
message StockInitialized {
  string product_id = 1;
  int32 quantity = 2;
  int32 low_stock_threshold = 3;
  google.protobuf.Timestamp initialized_at = 4;
}

message StockReceived {
  int32 quantity = 1;
  int32 new_on_hand = 2;
  string reference = 3;
  google.protobuf.Timestamp received_at = 4;
}

message StockReserved {
  int32 quantity = 1;
  string order_id = 2;
  int32 new_available = 3;
  google.protobuf.Timestamp reserved_at = 4;
  int32 new_reserved = 5;    // Total reserved after this event (fact, not delta)
  int32 new_on_hand = 6;     // Total on_hand after this event (unchanged by reserve)
}

message ReservationReleased {
  string order_id = 1;
  int32 quantity = 2;
  int32 new_available = 3;
  google.protobuf.Timestamp released_at = 4;
  int32 new_reserved = 5;    // Total reserved after this event (fact, not delta)
  int32 new_on_hand = 6;     // Total on_hand after this event (unchanged by release)
}

message ReservationCommitted {
  string order_id = 1;
  int32 quantity = 2;
  int32 new_on_hand = 3;
  google.protobuf.Timestamp committed_at = 4;
  int32 new_reserved = 5;    // Total reserved after this event (fact, not delta)
}

message LowStockAlert {
  string product_id = 1;
  int32 available = 2;
  int32 threshold = 3;
  google.protobuf.Timestamp alerted_at = 4;
}

// State (for snapshots)
message InventoryState {
  string product_id = 1;
  int32 on_hand = 2;
  int32 reserved = 3;
  int32 low_stock_threshold = 4;
  map<string, int32> reservations = 5;  // order_id -> quantity
}
